azure-iot-device/azure/iot/device/provisioning/pipeline/pipeline_stages_provisioning.py

```diff
@@ -44,3 +58,388 @@ def _run_op(self, op):
+                                    "Query Status operation returned a failed registration status  with a status code of {status_code}".format(
+                                        status_code=success_status_code
+                                    )
+                                )
+                        else:
+                            error = exceptions.ServiceError(
+                                "Query Status Operation encountered an invalid registration status {status} with a status code of {status_code}".format(
+                                    status=registration_status, status_code=success_status_code
+                                )
+                            )
+
+                        query_status_op.complete(error=error)
+
+                else:
+                    query_status_op.complete(error=error)
+
+            self.send_op_down(
+                pipeline_ops_base.RequestAndResponseOperation(
+                    request_type=constant.QUERY,
+                    method="GET",
+                    resource_location="/",
+                    query_params={"operation_id": op.operation_id},
+                    request_body=op.request_payload,
+                    callback=on_query_response,
+                )
+            )
+
+        else:
+            super(PollingStatusStage, self)._run_op(op)
+
+
+class RegistrationStage(CommonProvisioningStage):
+    """
+    This is the first stage that decides converts a registration request
+    into a normal request and response operation.
+    Upon the receipt of the response this stage decides whether
+    to send another registration request or send a query request.
+    Depending on the status and result of the response
+    this stage may also complete the registration process.
+    """
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _run_op(self, operation):
+        if isinstance(operation, pipeline_ops_provisioning.SendRegistrationRequestOperation):
+            initial_register_op = operation
+
+            def on_registration_response(op, error):
+                logger.debug(
+                    "{stage_name}({op_name}): Received response with status code {status_code} for SendRegistrationRequestOperation".format(
+                        stage_name=self.name, op_name=op.name, status_code=op.status_code
+                    )
+                )
+
+                # This could be an error that has been reported to this stage Or this
+                # could be an error because the service responded with status code >=300
+                error = self._get_error(op, REGISTER, error=error)
+
+                if not error:
+                    success_status_code = op.status_code
+                    retry_interval = (
+                        int(op.retry_after, 10)
+                        if op.retry_after is not None
+                        else constant.DEFAULT_POLLING_INTERVAL
+                    )
```

Comment:
This is only relevant in the >= 429 block

Reply:
I disagree.  If provisioning status == "assigning", then the client needs to retry regardless of the status_code.  

---
In reply to: [365408165](https://github.com/Azure/azure-iot-sdk-python/pull/432#discussion_r365408165) [](ancestors = 365408165)

Reply:
It's not used in the =="assigning" code block though. Are you saying it should be?

---
In reply to: [365977433](https://github.com/Azure/azure-iot-sdk-python/pull/432#discussion_r365977433) [](ancestors = 365977433,365408165)

Reply:
I don't know what really happens or what should happen, but It's possible that retry_after could be set for == "assigning".  

---
In reply to: [366026623](https://github.com/Azure/azure-iot-sdk-python/pull/432#discussion_r366026623) [](ancestors = 366026623,365977433,365408165)